Add `subject` to attestation manifests

relates to https://github.com/docker/build-push-action/issues/1260

`subject` has been added to the OCI spec 1.1.0: https://specs.opencontainers.org/image-spec/manifest/?v=v1.1.0#image-manifest-property-descriptions. This property specifies a [descriptor](https://specs.opencontainers.org/image-spec/descriptor/?v=v1.1.0) of another manifest.

Some registry providers like GAR use this property to link back to the original manifest. I think it would be fine to have this property attached to attestation manifests.

cc @wieringen @dvdksn @tianon

Tested here: https://github.com/docker/build-push-action/actions/runs/12182820627

Seems only Gitlab fails: https://github.com/docker/build-push-action/actions/runs/12182820627/job/33982750597#step:13:11

```
Exists    d9014c462484 hi.txt
Error response from registry: manifest blob unknown: blob unknown to registry: sha256:9b962b6abadee78b1d83d3b58a879fcbf0d66ea6d34c934ef3285301f3860b22
Error: Process completed with exit code 1.
```

![Image](https://github.com/user-attachments/assets/480f0fbe-df5d-46fe-8b56-ed6d246f0eb4)

Tested also locally:

```dockerfile
FROM alpine
RUN echo hello
```

```
$ docker buildx --builder builder build -t crazymax/buildkit-5561:multi --platform linux/amd64,linux/arm64 --push .
...
#10 exporting to image
#10 exporting layers done
#10 exporting manifest sha256:d3f28426b6941bb3d874b09d95019f4316284732cb5c8688b57d8159a28e0836 done
#10 exporting config sha256:4891e75c3b5e9ab03dbe24176ed3a580933fca9b47b44e2ad29338dd62cd05cd done
#10 exporting attestation manifest sha256:9b2bd7af0c3b6af640cea9cbd0998ade17afa04485810fa4720e40d466a68c8f
#10 exporting attestation manifest sha256:9b2bd7af0c3b6af640cea9cbd0998ade17afa04485810fa4720e40d466a68c8f 0.0s done
#10 exporting manifest sha256:9985f0b4525d4369e8f702daa8b96016cf62b94f07f0897b54a12c08565c2008 done
#10 exporting config sha256:dbaccb48e93c0be06f4e55ddab642cf7f8ff678f5f7c61209839ddbee07b099e done
#10 exporting attestation manifest sha256:04c144745d07d8a41a292144c9239c76f71f3487d121100797fb6c023c6d1826 0.0s done
#10 exporting manifest list sha256:166f78ab728dddc7f54a4cf12fdd37f604f4c1f12c521c0b78c29be6588752bf 0.0s done
#10 pushing layers
#10 ...

#11 [auth] crazymax/buildkit-5561:pull,push token for registry-1.docker.io
#11 DONE 0.0s

#10 exporting to image
#10 pushing layers 1.6s done
#10 pushing manifest for docker.io/crazymax/buildkit-5561:multi@sha256:166f78ab728dddc7f54a4cf12fdd37f604f4c1f12c521c0b78c29be6588752bf
#10 pushing manifest for docker.io/crazymax/buildkit-5561:multi@sha256:166f78ab728dddc7f54a4cf12fdd37f604f4c1f12c521c0b78c29be6588752bf 1.5s done
#10 DONE 3.3s
```

See https://explore.ggcr.dev/?image=crazymax/buildkit-5561:multi@sha256:166f78ab728dddc7f54a4cf12fdd37f604f4c1f12c521c0b78c29be6588752bf

![Image](https://github.com/user-attachments/assets/e2fc37c9-daef-4e99-8a59-9a89b3a47222)

```json
{
   "schemaVersion":2,
   "mediaType":"application/vnd.oci.image.index.v1+json",
   "manifests":[
      {
         "mediaType":"application/vnd.oci.image.manifest.v1+json",
         "size":667,
         "digest":"sha256:d3f28426b6941bb3d874b09d95019f4316284732cb5c8688b57d8159a28e0836",
         "platform":{
            "architecture":"amd64",
            "os":"linux"
         }
      },
      {
         "mediaType":"application/vnd.oci.image.manifest.v1+json",
         "size":667,
         "digest":"sha256:9985f0b4525d4369e8f702daa8b96016cf62b94f07f0897b54a12c08565c2008",
         "platform":{
            "architecture":"arm64",
            "os":"linux"
         }
      },
      {
         "mediaType":"application/vnd.oci.image.manifest.v1+json",
         "size":566,
         "digest":"sha256:9b2bd7af0c3b6af640cea9cbd0998ade17afa04485810fa4720e40d466a68c8f",
         "platform":{
            "architecture":"unknown",
            "os":"unknown"
         },
         "annotations":{
            "vnd.docker.reference.digest":"sha256:d3f28426b6941bb3d874b09d95019f4316284732cb5c8688b57d8159a28e0836",
            "vnd.docker.reference.type":"attestation-manifest"
         }
      },
      {
         "mediaType":"application/vnd.oci.image.manifest.v1+json",
         "size":566,
         "digest":"sha256:04c144745d07d8a41a292144c9239c76f71f3487d121100797fb6c023c6d1826",
         "platform":{
            "architecture":"unknown",
            "os":"unknown"
         },
         "annotations":{
            "vnd.docker.reference.digest":"sha256:9985f0b4525d4369e8f702daa8b96016cf62b94f07f0897b54a12c08565c2008",
            "vnd.docker.reference.type":"attestation-manifest"
         }
      }
   ]
}
```

Then attach an artifact with oras:

```
$ echo "Hello, world!" > hi.txt
$ oras attach --artifact-type doc/example docker.io/crazymax/buildkit-5561:multi@sha256:166f78ab728dddc7f54a4cf12fdd37f604f4c1f12c521c0b78c29be6588752bf hi.txt
✓ Exists    application/vnd.oci.empty.v1+json                                                                                                                                                         2/2  B 100.00%     0s
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Exists    hi.txt                                                                                                                                                                                  14/14  B 100.00%     0s
  └─ sha256:d9014c4624844aa5bac314773d6b689ad467fa4e1d1a50a1b8a99d5a95f72ff5
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                                                                                                            722/722  B 100.00%  632ms
  └─ sha256:712d3058686fa85a19077108e8a74affc44eee2c858144080f64b65e871fb57a
Attached to [registry] docker.io/crazymax/buildkit-5561:multi@sha256:166f78ab728dddc7f54a4cf12fdd37f604f4c1f12c521c0b78c29be6588752bf
Digest: sha256:712d3058686fa85a19077108e8a74affc44eee2c858144080f64b65e871fb57a
```

See "attached" manifest with subject attribute: https://explore.ggcr.dev/?image=crazymax%2Fbuildkit-5561%3Amulti%40sha256%3A712d3058686fa85a19077108e8a74affc44eee2c858144080f64b65e871fb57a

![Image](https://github.com/user-attachments/assets/5efad2a5-c54f-45ce-991c-176370a0a64e)

```json
{
   "schemaVersion":2,
   "mediaType":"application/vnd.oci.image.manifest.v1+json",
   "artifactType":"doc/example",
   "config":{
      "mediaType":"application/vnd.oci.empty.v1+json",
      "size":2,
      "digest":"sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "data":"e30="
   },
   "layers":[
      {
         "mediaType":"application/vnd.oci.image.layer.v1.tar",
         "size":14,
         "digest":"sha256:d9014c4624844aa5bac314773d6b689ad467fa4e1d1a50a1b8a99d5a95f72ff5",
         "annotations":{
            "org.opencontainers.image.title":"hi.txt"
         }
      }
   ],
   "subject":{
      "mediaType":"application/vnd.oci.image.index.v1+json",
      "size":1607,
      "digest":"sha256:166f78ab728dddc7f54a4cf12fdd37f604f4c1f12c521c0b78c29be6588752bf"
   },
   "annotations":{
      "org.opencontainers.image.created":"2024-12-05T15:37:17Z"
   }
}
```

--- a/README.md
+++ b/README.md
@@ -282,6 +282,7 @@ Keys supported by image output:
 * `push-by-digest=true`: push unnamed image
 * `registry.insecure=true`: push to insecure HTTP registry
 * `oci-mediatypes=true`: use OCI mediatypes in configuration JSON instead of Docker's
+* `oci-artifact=false`: use OCI artifact format for attestations
 * `unpack=true`: unpack image after creation (for use with containerd)
 * `dangling-name-prefix=<value>`: name image with `prefix@<digest>`, used for anonymous images
 * `name-canonical=true`: add additional canonical name `name@<digest>`

--- a/client/client_test.go
+++ b/client/client_test.go
@@ -192,7 +192,8 @@ var allTests = []func(t *testing.T, sb integration.Sandbox){
 	testPullWithLayerLimit,
 	testExportAnnotations,
 	testExportAnnotationsMediaTypes,
-	testExportAttestations,
+	testExportAttestationsOCIArtifact,
+	testExportAttestationsImageManifest,
 	testExportedImageLabels,
 	testAttestationDefaultSubject,
 	testSourceDateEpochLayerTimestamps,
@@ -8725,7 +8726,15 @@ func testExportAnnotationsMediaTypes(t *testing.T, sb integration.Sandbox) {
 	require.Equal(t, ocispecs.MediaTypeImageIndex, imgs2.Index.MediaType)
 }
 
-func testExportAttestations(t *testing.T, sb integration.Sandbox) {
+func testExportAttestationsOCIArtifact(t *testing.T, sb integration.Sandbox) {
+	testExportAttestations(t, sb, true)
+}
+
+func testExportAttestationsImageManifest(t *testing.T, sb integration.Sandbox) {
+	testExportAttestations(t, sb, false)
+}
+
+func testExportAttestations(t *testing.T, sb integration.Sandbox, ociArtifact bool) {
 	workers.CheckFeatureCompat(t, sb, workers.FeatureDirectPush)
 	requiresLinux(t)
 	c, err := New(sb.Context(), sb.Address())
@@ -8845,8 +8854,9 @@ func testExportAttestations(t *testing.T, sb integration.Sandbox) {
 				{
 					Type: ExporterImage,
 					Attrs: map[string]string{
-						"name": strings.Join(targets, ","),
-						"push": "true",
+						"name":         strings.Join(targets, ","),
+						"push":         "true",
+						"oci-artifact": strconv.FormatBool(ociArtifact),
 					},
 				},
 			},
@@ -8876,12 +8886,25 @@ func testExportAttestations(t *testing.T, sb integration.Sandbox) {
 		for i, att := range atts.Images {
 			require.Equal(t, ocispecs.MediaTypeImageManifest, att.Desc.MediaType)
 			require.Equal(t, "unknown/unknown", platforms.Format(*att.Desc.Platform))
-			require.Equal(t, "unknown/unknown", att.Img.OS+"/"+att.Img.Architecture)
 			require.Equal(t, attestation.DockerAnnotationReferenceTypeDefault, att.Desc.Annotations[attestation.DockerAnnotationReferenceType])
 			require.Equal(t, bases[i].Desc.Digest.String(), att.Desc.Annotations[attestation.DockerAnnotationReferenceDigest])
 			require.Equal(t, 2, len(att.Layers))
-			require.Equal(t, len(att.Layers), len(att.Img.RootFS.DiffIDs))
-			require.Equal(t, 0, len(att.Img.History))
+
+			if ociArtifact {
+				subject := att.Manifest.Subject
+				require.NotNil(t, subject)
+				require.Equal(t, bases[i].Desc, *subject)
+				require.Equal(t, "application/vnd.docker.attestation.manifest.v1+json", att.Manifest.ArtifactType)
+				require.Equal(t, ocispecs.DescriptorEmptyJSON, att.Manifest.Config)
+			} else {
+				require.Nil(t, att.Manifest.Subject)
+				require.Empty(t, att.Manifest.ArtifactType)
+
+				// image config is not included in the OCI artifact
+				require.Equal(t, "unknown/unknown", att.Img.OS+"/"+att.Img.Architecture)
+				require.Equal(t, len(att.Layers), len(att.Img.RootFS.DiffIDs))
+				require.Equal(t, 0, len(att.Img.History))
+			}
 
 			var attest intoto.Statement
 			require.NoError(t, json.Unmarshal(att.LayersRaw[0], &attest))

--- a/exporter/containerimage/exptypes/keys.go
+++ b/exporter/containerimage/exptypes/keys.go
@@ -44,6 +44,9 @@ var (
 	// Value: bool <true|false>
 	OptKeyOCITypes ImageExporterOptKey = "oci-mediatypes"
 
+	// Use OCI artifact format for the attestation manifest.
+	OptKeyOCIArtifact ImageExporterOptKey = "oci-artifact"
+
 	// Force attestation to be attached.
 	// Value: bool <true|false>
 	OptKeyForceInlineAttestations ImageExporterOptKey = "attestation-inline"

--- a/exporter/containerimage/opts.go
+++ b/exporter/containerimage/opts.go
@@ -17,6 +17,7 @@ type ImageCommitOpts struct {
 	ImageName   string
 	RefCfg      cacheconfig.RefConfig
 	OCITypes    bool
+	OCIArtifact bool
 	Annotations AnnotationsGroup
 	Epoch       *time.Time
 
@@ -49,6 +50,8 @@ func (c *ImageCommitOpts) Load(ctx context.Context, opt map[string]string) (map[
 			c.ImageName = v
 		case exptypes.OptKeyOCITypes:
 			err = parseBoolWithDefault(&c.OCITypes, k, v, true)
+		case exptypes.OptKeyOCIArtifact:
+			err = parseBool(&c.OCIArtifact, k, v)
 		case exptypes.OptKeyForceInlineAttestations:
 			err = parseBool(&c.ForceInlineAttestations, k, v)
 		case exptypes.OptKeyPreferNondistLayers:
@@ -66,6 +69,9 @@ func (c *ImageCommitOpts) Load(ctx context.Context, opt map[string]string) (map[
 
 	if c.RefCfg.Compression.Type.OnlySupportOCITypes() {
 		c.EnableOCITypes(ctx, c.RefCfg.Compression.Type.String())
+	}
+	if c.OCIArtifact && !c.OCITypes {
+		c.EnableOCITypes(ctx, "oci-artifact")
 	}
 
 	c.Annotations = c.Annotations.Merge(as)

--- a/exporter/containerimage/writer.go
+++ b/exporter/containerimage/writer.go
@@ -46,6 +46,8 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
+const attestationManifestArtifactType = "application/vnd.docker.attestation.manifest.v1+json"
+
 type WriterOpt struct {
 	Snapshotter  snapshot.Snapshotter
 	ContentStore content.Store
@@ -312,7 +314,7 @@ func (ic *ImageWriter) Commit(ctx context.Context, inp *exporter.Source, session
 				return nil, err
 			}
 
-			desc, err := ic.commitAttestationsManifest(ctx, opts, desc.Digest.String(), stmts)
+			desc, err := ic.commitAttestationsManifest(ctx, opts, *desc, stmts, opts.OCIArtifact)
 			if err != nil {
 				return nil, err
 			}
@@ -553,7 +555,7 @@ func (ic *ImageWriter) commitDistributionManifest(ctx context.Context, opts *Ima
 	}, &configDesc, nil
 }
 
-func (ic *ImageWriter) commitAttestationsManifest(ctx context.Context, opts *ImageCommitOpts, target string, statements []intoto.Statement) (*ocispecs.Descriptor, error) {
+func (ic *ImageWriter) commitAttestationsManifest(ctx context.Context, opts *ImageCommitOpts, target ocispecs.Descriptor, statements []intoto.Statement, ociArtifact bool) (*ocispecs.Descriptor, error) {
 	var (
 		manifestType = ocispecs.MediaTypeImageManifest
 		configType   = ocispecs.MediaTypeImageConfig
@@ -588,15 +590,21 @@ func (ic *ImageWriter) commitAttestationsManifest(ctx context.Context, opts *Ima
 		layers[i] = desc
 	}
 
-	config, err := attestationsConfig(layers)
-	if err != nil {
-		return nil, err
-	}
-	configDigest := digest.FromBytes(config)
-	configDesc := ocispecs.Descriptor{
-		Digest:    configDigest,
-		Size:      int64(len(config)),
-		MediaType: configType,
+	configDesc := ocispecs.DescriptorEmptyJSON
+	config := configDesc.Data
+
+	if !ociArtifact {
+		var err error
+		config, err = attestationsConfig(layers)
+		if err != nil {
+			return nil, err
+		}
+		configDigest := digest.FromBytes(config)
+		configDesc = ocispecs.Descriptor{
+			Digest:    configDigest,
+			Size:      int64(len(config)),
+			MediaType: configType,
+		}
 	}
 
 	mfst := ocispecs.Manifest{
@@ -604,15 +612,16 @@ func (ic *ImageWriter) commitAttestationsManifest(ctx context.Context, opts *Ima
 		Versioned: specs.Versioned{
 			SchemaVersion: 2,
 		},
-		Config: ocispecs.Descriptor{
-			Digest:    configDigest,
-			Size:      int64(len(config)),
-			MediaType: configType,
-		},
+		Config: configDesc,
+	}
+
+	if ociArtifact {
+		mfst.ArtifactType = attestationManifestArtifactType
+		mfst.Subject = &target
 	}
 
 	labels := map[string]string{
-		"containerd.io/gc.ref.content.0": configDigest.String(),
+		"containerd.io/gc.ref.content.0": configDesc.Digest.String(),
 	}
 	for i, desc := range layers {
 		desc.Annotations = RemoveInternalLayerAnnotations(desc.Annotations, opts.OCITypes)
@@ -635,7 +644,7 @@ func (ic *ImageWriter) commitAttestationsManifest(ctx context.Context, opts *Ima
 	if err := content.WriteBlob(ctx, ic.opt.ContentStore, mfstDigest.String(), bytes.NewReader(mfstJSON), mfstDesc, content.WithLabels((labels))); err != nil {
 		return nil, done(errors.Wrapf(err, "error writing manifest blob %s", mfstDigest))
 	}
-	if err := content.WriteBlob(ctx, ic.opt.ContentStore, configDigest.String(), bytes.NewReader(config), configDesc); err != nil {
+	if err := content.WriteBlob(ctx, ic.opt.ContentStore, configDesc.Digest.String(), bytes.NewReader(config), configDesc); err != nil {
 		return nil, done(errors.Wrap(err, "error writing config blob"))
 	}
 	done(nil)
@@ -646,7 +655,7 @@ func (ic *ImageWriter) commitAttestationsManifest(ctx context.Context, opts *Ima
 		MediaType: manifestType,
 		Annotations: map[string]string{
 			attestationTypes.DockerAnnotationReferenceType:   attestationTypes.DockerAnnotationReferenceTypeDefault,
-			attestationTypes.DockerAnnotationReferenceDigest: target,
+			attestationTypes.DockerAnnotationReferenceDigest: string(target.Digest),
 		},
 	}, nil
 }

